File: src/commands/domain.rs
use crate::api::{self, Client};
use crate::commands::credit::{format_date, format_idr};
use crate::output::*;

pub fn cmd_domain(args: &[String]) -> Result<(), String> {
    if args.is_empty() || args[0] == "--help" || args[0] == "-h" {
        crate::commands::help::print_domain_help();
        return Ok(());
    }

    match args[0].as_str() {
        "enable" => {
            if args.len() < 2 {
                print_error("Usage: dalang domain enable <vps-name>");
                return Err("missing VPS name".to_string());
            }
            domain_enable(&args[1])
        }
        "list" | "ls" => {
            if args.len() < 2 {
                print_error("Usage: dalang domain list <vps-name>");
                return Err("missing VPS name".to_string());
            }
            domain_list(&args[1])
        }
        "add" => {
            if args.len() < 3 {
                print_error("Usage: dalang domain add <vps-name> <domain>");
                return Err("missing VPS name or domain".to_string());
            }
            domain_add(&args[1], &args[2])
        }
        "verify" => {
            if args.len() < 2 {
                print_error("Usage: dalang domain verify <domain>");
                return Err("missing domain".to_string());
            }
            domain_verify(&args[1])
        }
        "remove" | "rm" => {
            if args.len() < 2 {
                print_error("Usage: dalang domain remove <domain>");
                return Err("missing domain".to_string());
            }
            domain_remove(&args[1])
        }
        other => {
            print_error(&format!("Unknown domain subcommand: {}", other));
            Err(format!("unknown subcommand: {}", other))
        }
    }
}

fn find_vps_by_name(client: &Client, name: &str) -> Result<api::VPS, String> {
    let resp = client
        .get("/vps/list")
        .map_err(|e| format!("failed to fetch VPS list: {}", e))?;

    let vps_data: api::VPSListResponse =
        serde_json::from_slice(&resp).map_err(|e| format!("failed to parse response: {}", e))?;

    for v in vps_data.data {
        if v.name == name || v.display_name == name {
            return Ok(v);
        }
    }

    Err(format!("VPS '{}' not found", name))
}

fn domain_enable(vps_name: &str) -> Result<(), String> {
    let client = Client::new_authenticated()?;
    let vps = find_vps_by_name(&client, vps_name)?;

    if vps.custom_domain_enabled == 1 {
        print_success(&format!(
            "Custom domain is already enabled for {}",
            vps_name
        ));
        return Ok(());
    }

    if !yes_flag() {
        println!(
            "\nEnable custom domain for {}{}{}?",
            CYAN, vps_name, RESET
        );
        println!(
            "Price: {}{}/month{}\n",
            YELLOW,
            format_idr(vps.custom_domain_price as i64),
            RESET
        );
        if !confirm_prompt("Proceed with payment?") {
            print_info("Cancelled");
            return Ok(());
        }
    }

    print_info("Creating invoice...");

    let body = serde_json::json!({"vps_id": vps.id});
    let resp = client
        .post("/vps/addon/custom-domain", Some(&body))
        .map_err(|e| format!("failed to enable custom domain: {}", e))?;

    let result: api::DomainEnableResponse = serde_json::from_slice(&resp)
        .map_err(|e| format!("failed to parse response: {}", e))?;

    if !result.success {
        print_error(&result.message);
        return Err(result.message.clone());
    }

    if json_output() {
        println!("{}", String::from_utf8_lossy(&resp));
        return Ok(());
    }

    print_success("Invoice created!");
    println!(
        "\n  Amount: {}",
        format_idr(result.data.amount)
    );
    println!(
        "  Pay at: {}{}{}\n",
        CYAN, result.data.invoice_url, RESET
    );
    print_info("After payment, custom domain will be automatically enabled");

    Ok(())
}

fn domain_list(vps_name: &str) -> Result<(), String> {
    let client = Client::new_authenticated()?;
    let vps = find_vps_by_name(&client, vps_name)?;

    if vps.custom_domain_enabled != 1 {
        print_warn(&format!(
            "Custom domain is not enabled for {}",
            vps_name
        ));
        println!(
            "\nRun '{}dalang domain enable {}{}' to activate",
            CYAN, vps_name, RESET
        );
        return Ok(());
    }

    let resp = client
        .get(&format!("/vps/domains/list?vps_id={}", vps.id))
        .map_err(|e| format!("failed to list domains: {}", e))?;

    let result: api::DomainListResponse = serde_json::from_slice(&resp)
        .map_err(|e| format!("failed to parse response: {}", e))?;

    if json_output() {
        println!("{}", String::from_utf8_lossy(&resp));
        return Ok(());
    }

    println!(
        "\n{}Custom Domains for {}{}",
        BOLD, vps_name, RESET
    );
    println!("{}", "\u{2500}".repeat(60));

    if result.data.is_empty() {
        print_info("No custom domains configured");
        println!(
            "\nAdd a domain with: {}dalang domain add {} <your-domain.com>{}",
            CYAN, vps_name, RESET
        );
        return Ok(());
    }

    println!("{:<30} {:<12} {}", "DOMAIN", "STATUS", "CREATED");
    println!("{}", "\u{2500}".repeat(60));

    for d in &result.data {
        let status_color = match d.status.as_str() {
            "active" => GREEN,
            "failed" => RED,
            _ => YELLOW,
        };

        println!(
            "{:<30} {}{:<12}{} {}",
            d.domain,
            status_color,
            d.status,
            RESET,
            format_date(&d.created_at),
        );
    }
    println!();

    Ok(())
}

fn domain_add(vps_name: &str, domain: &str) -> Result<(), String> {
    let client = Client::new_authenticated()?;
    let vps = find_vps_by_name(&client, vps_name)?;

    if vps.custom_domain_enabled != 1 {
        print_error(&format!(
            "Custom domain is not enabled for {}",
            vps_name
        ));
        println!(
            "\nRun '{}dalang domain enable {}{}' first",
            CYAN, vps_name, RESET
        );
        return Err("custom domain not enabled".to_string());
    }

    print_info(&format!("Adding domain {}...", domain));

    let body = serde_json::json!({
        "vps_id": vps.id,
        "domain": domain,
    });

    let resp = client
        .post("/vps/domains/add", Some(&body))
        .map_err(|e| format!("failed to add domain: {}", e))?;

    let result: api::DomainAddResponse = serde_json::from_slice(&resp)
        .map_err(|e| format!("failed to parse response: {}", e))?;

    if !result.success {
        print_error(&result.message);
        return Err(result.message.clone());
    }

    if json_output() {
        println!("{}", String::from_utf8_lossy(&resp));
        return Ok(());
    }

    print_success("Domain added!");
    println!(
        "\n{}Configure these DNS records:{}\n",
        BOLD, RESET
    );

    for record in &result.data.dns_records {
        let rtype = record.get("type").map(|s| s.as_str()).unwrap_or("");
        let rname = record.get("name").map(|s| s.as_str()).unwrap_or("");
        let rvalue = record.get("value").map(|s| s.as_str()).unwrap_or("");
        let rdesc = record.get("description").map(|s| s.as_str()).unwrap_or("");

        println!("  {}{:<6}{}  {}", CYAN, rtype, RESET, rname);
        println!("          \u{2192} {}", rvalue);
        println!("          {}{}{}\n", YELLOW, rdesc, RESET);
    }

    println!(
        "After configuring DNS, run: {}dalang domain verify {}{}\n",
        CYAN, domain, RESET
    );

    Ok(())
}

fn domain_verify(domain: &str) -> Result<(), String> {
    let client = Client::new_authenticated()?;

    print_info(&format!("Verifying domain {}...", domain));

    let body = serde_json::json!({"domain": domain});
    let resp = client
        .post("/vps/domains/verify", Some(&body))
        .map_err(|e| format!("failed to verify domain: {}", e))?;

    let result: api::DomainVerifyResponse = serde_json::from_slice(&resp)
        .map_err(|e| format!("failed to parse response: {}", e))?;

    if json_output() {
        println!("{}", String::from_utf8_lossy(&resp));
        return Ok(());
    }

    if result.success && result.data.status == "active" {
        print_success(&format!("Domain {} is verified and active!", domain));
        return Ok(());
    }

    print_warn("Domain verification in progress");
    println!("\n  Domain: {}", result.data.domain);
    println!("  Status: {}", result.data.status);
    println!("  SSL:    {}\n", result.data.ssl_status);

    if !result.data.dns_records.is_empty() {
        println!(
            "{}Ensure these DNS records are configured:{}\n",
            BOLD, RESET
        );
        for record in &result.data.dns_records {
            let rtype = record.get("type").map(|s| s.as_str()).unwrap_or("");
            let rname = record.get("name").map(|s| s.as_str()).unwrap_or("");
            let rvalue = record.get("value").map(|s| s.as_str()).unwrap_or("");

            println!("  {}{:<6}{}  {}", CYAN, rtype, RESET, rname);
            println!("          \u{2192} {}\n", rvalue);
        }
    }

    print_info("DNS propagation can take up to 24 hours. Try again later.");

    Ok(())
}

fn domain_remove(domain: &str) -> Result<(), String> {
    let client = Client::new_authenticated()?;

    if !yes_flag() {
        println!("\nRemove domain {}{}{}?\n", CYAN, domain, RESET);
        if !confirm_prompt("Are you sure?") {
            print_info("Cancelled");
            return Ok(());
        }
    }

    print_info(&format!("Removing domain {}...", domain));

    let body = serde_json::json!({"domain": domain});
    let resp = client
        .post("/vps/domains/remove", Some(&body))
        .map_err(|e| format!("failed to remove domain: {}", e))?;

    let result: api::GenericResponse = serde_json::from_slice(&resp)
        .map_err(|e| format!("failed to parse response: {}", e))?;

    if !result.success {
        print_error(&result.message);
        return Err(result.message.clone());
    }

    print_success("Domain removed successfully");
    Ok(())
}

File: cmd/domain.go
package cmd

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"

	"github.com/dalang-io/dalang-cli/internal/api"
)

func cmdDomain(args []string) error {
	if len(args) == 0 || args[0] == "--help" || args[0] == "-h" {
		printDomainHelp()
		return nil
	}

	switch args[0] {
	case "enable":
		if len(args) < 2 {
			printUsage("Usage: dalang domain enable <vps-name>")
			return fmt.Errorf("missing VPS name")
		}
		return domainEnable(args[1])
	case "list", "ls":
		if len(args) < 2 {
			printUsage("Usage: dalang domain list <vps-name>")
			return fmt.Errorf("missing VPS name")
		}
		return domainList(args[1])
	case "add":
		if len(args) < 3 {
			printUsage("Usage: dalang domain add <vps-name> <domain>")
			return fmt.Errorf("missing VPS name or domain")
		}
		return domainAdd(args[1], args[2])
	case "verify":
		if len(args) < 2 {
			printUsage("Usage: dalang domain verify <domain>")
			return fmt.Errorf("missing domain")
		}
		return domainVerify(args[1])
	case "remove", "rm":
		if len(args) < 2 {
			printUsage("Usage: dalang domain remove <domain>")
			return fmt.Errorf("missing domain")
		}
		return domainRemove(args[1])
	default:
		return fmt.Errorf("unknown domain subcommand: %s", args[0])
	}
}

func domainEnable(vpsName string) error {
	client, err := api.NewAuthenticatedClient()
	if err != nil {
		return err
	}
	client.Verbose = VerboseOutput

	// Find VPS by name
	vps, err := findVPSByName(client, vpsName)
	if err != nil {
		return err
	}

	if vps.CustomDomainEnabled == 1 {
		printSuccess("Custom domain is already enabled for %s", vpsName)
		return nil
	}

	if !yesFlag {
		fmt.Printf("\nEnable custom domain for %s%s%s?\n", colorCyan, vpsName, colorReset)
		fmt.Printf("Price: %s%s/month%s\n\n", colorYellow, formatIDR(int64(vps.CustomDomainPrice)), colorReset)
		if !confirmPrompt("Proceed with payment?") {
			printInfo("Cancelled")
			return nil
		}
	}

	printInfo("Creating invoice...")

	resp, err := client.Post("/vps/addon/custom-domain", map[string]interface{}{
		"vps_id": vps.ID,
	})
	if err != nil {
		return fmt.Errorf("failed to enable custom domain: %w", err)
	}

	var result struct {
		Success bool   `json:"success"`
		Message string `json:"message"`
		Data    struct {
			InvoiceURL string `json:"invoice_url"`
			Amount     int    `json:"amount"`
		} `json:"data"`
	}

	if err := json.Unmarshal(resp, &result); err != nil {
		return fmt.Errorf("failed to parse response: %w", err)
	}

	if !result.Success {
		return errors.New(result.Message)
	}

	if jsonOutput {
		fmt.Println(string(resp))
		return nil
	}

	printSuccess("Invoice created!")
	fmt.Printf("\n  Amount: %s\n", formatIDR(int64(result.Data.Amount)))
	fmt.Printf("  Pay at: %s%s%s\n\n", colorCyan, result.Data.InvoiceURL, colorReset)
	printInfo("After payment, custom domain will be automatically enabled")

	return nil
}

func domainList(vpsName string) error {
	client, err := api.NewAuthenticatedClient()
	if err != nil {
		return err
	}
	client.Verbose = VerboseOutput

	// Find VPS by name
	vps, err := findVPSByName(client, vpsName)
	if err != nil {
		return err
	}

	if vps.CustomDomainEnabled != 1 {
		printWarn("Custom domain is not enabled for %s", vpsName)
		fmt.Printf("\nRun '%sdalang domain enable %s%s' to activate\n", colorCyan, vpsName, colorReset)
		return nil
	}

	resp, err := client.Get(fmt.Sprintf("/vps/domains/list?vps_id=%s", vps.ID))
	if err != nil {
		return fmt.Errorf("failed to list domains: %w", err)
	}

	var result struct {
		Success bool `json:"success"`
		Data    []struct {
			ID        int    `json:"id"`
			Domain    string `json:"domain"`
			Status    string `json:"status"`
			CreatedAt string `json:"created_at"`
		} `json:"data"`
	}

	if err := json.Unmarshal(resp, &result); err != nil {
		return fmt.Errorf("failed to parse response: %w", err)
	}

	if jsonOutput {
		fmt.Println(string(resp))
		return nil
	}

	fmt.Printf("\n%sCustom Domains for %s%s\n", colorBold, vpsName, colorReset)
	fmt.Println(strings.Repeat("─", 60))

	if len(result.Data) == 0 {
		printInfo("No custom domains configured")
		fmt.Printf("\nAdd a domain with: %sdalang domain add %s <your-domain.com>%s\n", colorCyan, vpsName, colorReset)
		return nil
	}

	fmt.Printf("%-30s %-12s %s\n", "DOMAIN", "STATUS", "CREATED")
	fmt.Println(strings.Repeat("─", 60))

	for _, d := range result.Data {
		statusColor := colorYellow
		if d.Status == "active" {
			statusColor = colorGreen
		} else if d.Status == "failed" {
			statusColor = colorRed
		}

		fmt.Printf("%-30s %s%-12s%s %s\n",
			d.Domain,
			statusColor, d.Status, colorReset,
			formatDate(d.CreatedAt),
		)
	}
	fmt.Println()

	return nil
}

func domainAdd(vpsName, domain string) error {
	client, err := api.NewAuthenticatedClient()
	if err != nil {
		return err
	}
	client.Verbose = VerboseOutput

	// Find VPS by name
	vps, err := findVPSByName(client, vpsName)
	if err != nil {
		return err
	}

	if vps.CustomDomainEnabled != 1 {
		fmt.Printf("Run '%sdalang domain enable %s%s' first\n", colorCyan, vpsName, colorReset)
		return fmt.Errorf("custom domain is not enabled for %s", vpsName)
	}

	printInfo("Adding domain %s...", domain)

	resp, err := client.Post("/vps/domains/add", map[string]interface{}{
		"vps_id": vps.ID,
		"domain": domain,
	})
	if err != nil {
		return fmt.Errorf("failed to add domain: %w", err)
	}

	var result struct {
		Success bool   `json:"success"`
		Message string `json:"message"`
		Data    struct {
			Domain       string              `json:"domain"`
			Status       string              `json:"status"`
			DNSRecords   []map[string]string `json:"dns_records"`
			Instructions string              `json:"instructions"`
		} `json:"data"`
	}

	if err := json.Unmarshal(resp, &result); err != nil {
		return fmt.Errorf("failed to parse response: %w", err)
	}

	if !result.Success {
		return errors.New(result.Message)
	}

	if jsonOutput {
		fmt.Println(string(resp))
		return nil
	}

	printSuccess("Domain added!")
	fmt.Printf("\n%sConfigure these DNS records:%s\n\n", colorBold, colorReset)

	for _, record := range result.Data.DNSRecords {
		fmt.Printf("  %s%-6s%s  %s\n", colorCyan, record["type"], colorReset, record["name"])
		fmt.Printf("          → %s\n", record["value"])
		fmt.Printf("          %s%s%s\n\n", colorYellow, record["description"], colorReset)
	}

	fmt.Printf("After configuring DNS, run: %sdalang domain verify %s%s\n\n", colorCyan, domain, colorReset)

	return nil
}

func domainVerify(domain string) error {
	client, err := api.NewAuthenticatedClient()
	if err != nil {
		return err
	}
	client.Verbose = VerboseOutput

	printInfo("Verifying domain %s...", domain)

	resp, err := client.Post("/vps/domains/verify", map[string]interface{}{
		"domain": domain,
	})
	if err != nil {
		return fmt.Errorf("failed to verify domain: %w", err)
	}

	var result struct {
		Success bool   `json:"success"`
		Message string `json:"message"`
		Data    struct {
			Domain     string              `json:"domain"`
			Status     string              `json:"status"`
			SSLStatus  string              `json:"ssl_status"`
			DNSRecords []map[string]string `json:"dns_records"`
		} `json:"data"`
	}

	if err := json.Unmarshal(resp, &result); err != nil {
		return fmt.Errorf("failed to parse response: %w", err)
	}

	if jsonOutput {
		fmt.Println(string(resp))
		return nil
	}

	if result.Success && result.Data.Status == "active" {
		printSuccess("Domain %s is verified and active!", domain)
		return nil
	}

	printWarn("Domain verification in progress")
	fmt.Printf("\n  Domain: %s\n", result.Data.Domain)
	fmt.Printf("  Status: %s\n", result.Data.Status)
	fmt.Printf("  SSL:    %s\n\n", result.Data.SSLStatus)

	if len(result.Data.DNSRecords) > 0 {
		fmt.Printf("%sEnsure these DNS records are configured:%s\n\n", colorBold, colorReset)
		for _, record := range result.Data.DNSRecords {
			fmt.Printf("  %s%-6s%s  %s\n", colorCyan, record["type"], colorReset, record["name"])
			fmt.Printf("          → %s\n\n", record["value"])
		}
	}

	printInfo("DNS propagation can take up to 24 hours. Try again later.")

	return nil
}

func domainRemove(domain string) error {
	client, err := api.NewAuthenticatedClient()
	if err != nil {
		return err
	}
	client.Verbose = VerboseOutput

	if !yesFlag {
		fmt.Printf("\nRemove domain %s%s%s?\n\n", colorCyan, domain, colorReset)
		if !confirmPrompt("Are you sure?") {
			printInfo("Cancelled")
			return nil
		}
	}

	printInfo("Removing domain %s...", domain)

	resp, err := client.Post("/vps/domains/remove", map[string]interface{}{
		"domain": domain,
	})
	if err != nil {
		return fmt.Errorf("failed to remove domain: %w", err)
	}

	var result struct {
		Success bool   `json:"success"`
		Message string `json:"message"`
	}

	if err := json.Unmarshal(resp, &result); err != nil {
		return fmt.Errorf("failed to parse response: %w", err)
	}

	if !result.Success {
		return errors.New(result.Message)
	}

	printSuccess("Domain removed successfully")
	return nil
}

func printDomainHelp() {
	fmt.Printf(`%sdalang domain%s - Manage custom domains

%sUSAGE:%s
    dalang domain enable <vps-name>           Enable custom domain (+10K/month)
    dalang domain list <vps-name>             List custom domains
    dalang domain add <vps-name> <domain>     Add a custom domain
    dalang domain verify <domain>             Verify domain DNS configuration
    dalang domain remove <domain>             Remove a custom domain

%sDESCRIPTION:%s
    Manage custom domains for your VPS. First enable the custom domain
    addon, then add and verify your domains.

%sEXAMPLES:%s
    dalang domain enable MyVM                 # Enable custom domain feature
    dalang domain add MyVM example.com        # Add a domain
    dalang domain verify example.com          # Verify DNS is configured
    dalang domain list MyVM                   # List all domains
    dalang domain remove example.com          # Remove a domain

%sWORKFLOW:%s
    1. Enable custom domain: dalang domain enable <vps>
    2. Pay the invoice
    3. Add your domain: dalang domain add <vps> <domain>
    4. Configure DNS records as shown
    5. Verify: dalang domain verify <domain>
`,
		colorCyan, colorReset,
		colorYellow, colorReset,
		colorYellow, colorReset,
		colorYellow, colorReset,
		colorYellow, colorReset,
	)
}
